core/trino-main/src/main/java/io/trino/connector/ConnectorContextInstance.java

```diff
@@ -30,6 +31,8 @@
 public class ConnectorContextInstance
         implements ConnectorContext
 {
+    private static final Optional<String> SPI_VERSION = Optional.ofNullable(ConnectorContext.class.getPackage().getImplementationVersion());
```

Comment:
An alternative would be to start using SemVer for SPI and only check for major version compatibility.

Reply:
> any other community plugins will have issues following this cadence of releases.

A plugin maintained outside of Trino does not have to add such checks.

> Instead of making people upgrade faster, it may make them stay on older versions for longer.

@hashhar please explain what you're concerned about

> An alternative would be to start using SemVer for SPI and only check for major version compatibility.

We consciously do not use semantic versioning. Or, actually, we use it: every version is a major version.

Reply:
As long as the engine doesn't start verifying the SPI version I'm fine with this. But this PR suggests we may add engine checks too in future. In current shape it's a good change for my opinion.

Reply:
If we don't use Semantic Versioning, the version number has no meaning, and shouldn't be used for a version check.

All SEP connectors built on `trino-base-jdbc` are broken by the addition of this check, and the workaround involves delegating to the `JdbcConnectorFactory(...)` from the Trino release, which (of course) will pass the Versions check.

Reply:
each version is a major version with potentially breaking changes - i.e. all versions are X.0.0

Semantic makes sense for libraries not as much for applications since defining breaking changes is very gray area for applications.

Reply:
Re: Semantic versions - you are correct. We treat them all as major. I guess my point was more to the fact that the SPI-"version" is being checked against (potentially) non-Trino connector's version, and in cases like these, the version check is not what we want - we want a SPI-level check, not a Connector-implementation-jar level check:

I have a plugin, `com.contoso:contoso-plugin:371` (version is completely independent of Trino's, in this example they just happen to match...).

It has the following dependencies in the effective dependency tree:
```
        <dependency>
            <groupId>io.trino</groupId>
            <artifactId>trino-spi</artifactId>
            <version>371</version>
        </dependency>

        <dependency>
            <groupId>io.trino</groupId>
            <artifactId>trino-base-jdbc</artifactId>
            <version>300-SNAPSHOT</version>
        </dependency>

```

Today, `Versions.java` will allow us to subclass `JdbcConnectorFactory` from `io.trino:trino-base-jdbc:300-SNAPSHOT` and it will succeed because it is checking `com.contoso:contoso-plugin:371`'s version number and matching it to `io.trino:trino-spi:371`'s `io/trino/spi/trino-spi-version.txt` version. 
I would suggest `Versions.java` should walk the class hierarchy to find if any `io.trino` classes are involved, and only check them (or we could add an `spi-version:` attribute to artifact metadata and compare based on that (if present)).